### PR TITLE
Dynamic Sizing Viewport

### PR DESCRIPTION
**Pull request by nevimer**
## About The Pull Request

Redo of my old PR to here & TG, of a port of Nebula's dynamic viewport.  What does it do? It allows you to resize your viewport to server-side limits, providing a dynamic and screen-filling experience. 
See this for a demo.
https://renny.streamasaurus.com/index.php/s/s76ZjNEpLYr2fJS

I'll fill this bit out a bit more later, but, see my old PR's for a better description.

## Why It's Good For The Game

Brings the game experience a bit more up to modern times for players. 

## Changelog
:cl:
add: Dynamic scaling viewport
refactor: Screen resizes are now dynamically scaling and make use of the entire screen space if possible. 
/:cl:

### CODE REVIEW

**Comment by Zandario** [comment]
Blessed

**Comment by nevimer** [comment]
Good for another TM

**Comment by Knightfall5** [comment]
Chat bar is way too small if I want the entire screen to fit using stretch to fit
https://i.imgur.com/bOYVcMl.png
For comparison, this is what widescreen looks like on TGMC using the same settings
https://i.imgur.com/g9mGAeA.png
Still a bit small for my tastes, but usable atleast

**Comment by nevimer** [comment]
I'll add more precise options for scaling. Stretch to fit has a few issues

**Comment by nevimer** [comment]
> 



> Chat bar is way too small if I want the entire screen to fit using stretch to fit https://i.imgur.com/bOYVcMl.png For comparison, this is what widescreen looks like on TGMC using the same settings https://i.imgur.com/g9mGAeA.png Still a bit small for my tastes, but usable atleast

oof, yeah, 16:10 ratios have a hard time with anything. Try dragging your chat wider, The intent of the scaling is for YOU, the player, to decide how large your chat should be.

File: code/controllers/configuration/entries/game_options.dm [hunk 1]
```diff
@@ -51,3 +51,13 @@
 
 /datum/config_entry/flag/emojis
 	default = TRUE
+
+/datum/config_entry/number/max_client_view_x
+	default = 21
```
Comment (moonheart08) [inline code/controllers/configuration/entries/game_options.dm:56]: you should make this 29, it'd be funny

**Review by Zandario — DISMISSED** [review]
Looking good, not sure what issues are left, though this is good enough for me now.

**Comment by nevimer** [comment]
I intend to finish this once I am home tonight, and if I don't finish it tonight, tomorrow. Just need to do silicons locking suggestions from discord and investigate a pref issue not loading right.

File: code/modules/client/viewport.dm [hunk 2]
```diff
@@ -0,0 +1,85 @@
+GLOBAL_VAR_INIT(lock_client_view_x, null)
+GLOBAL_VAR_INIT(lock_client_view_y, null)
+GLOBAL_LIST_INIT(valid_icon_sizes, list(32, 48, 64, 72, 96, 128))
+
+/datum/preferences
+	var/icon_size = 64
+
+/client
+	var/last_view_x_dim = 7
+	var/last_view_y_dim = 7
```
Comment (Zandario) [inline code/modules/client/viewport.dm:10]: I'd prefer these to not be isolated from the rest of the pref & client var defines THOUGH this is fine for now as we still need to go through with our preference refactor

**Review by Zandario — APPROVED** [review]

**Pull request merged**